mipengine/controller/controller.py

```diff
@@ -260,8 +259,10 @@ async def exec_algorithm(
         except Exception as exc:
             algo_execution_logger.error(traceback.format_exc())
             raise exc
+
         finally:
-            self._cleaner.release_context_id(context_id=context_id)
+            if not self._cleaner.cleanup_context_id(context_id=context_id):
```

Comment:
I think its better for the controller to not know about the logic of the cleaner.

You can move the if in line 264 at the end of the .cleanup_context_id.
controller line:264  Could be: `self._cleaner.cleanup_context_id(context_id=context_id)`
cleaner line: 96 Could be: 
`if not self._exec_cleanup(entry):`
`    self._cleaner.release_context_id(context_id=context_id)`

Reply:
the controller "knowing" that there are 2 mechanisms for triggering the cleanup, a synchronous and an asynchronous, is intended. Doing what you propose, inside cleaner.py, first you mix the 2 mechanisms and second it might lead to undefined behavior since the 2 classes in the cleaner.py module are not thread safe.
Essentially this would add much complexity without much gain on any other aspect.

Reply:
Controller should only say to the cleaner please  'cleanup' nothing else.
Everything else is for cleaner to decide this is the correct way imo.
Like I said these are suggestions, I do not have a strong opinion on them.
It is already approved by me so you can change it according to my suggestion or merge it as is both work for me.

Reply:
"Controller should only say to the cleaner please 'cleanup' nothing else.
Everything else is for cleaner to decide this is the correct way imo."

This mechanism you describe is how the cleanup procedure was implemented before the current change.
1. The `Controller` would first call `add_contextid_for_cleanup(<context_id,node_ids>)` which would "inform" the `Cleaner` that db artifacts with a specific `context_id` will be (potentially) created on the specified nodes.
2. Then when these artifacts are not needed anymore, because the algorithm execution, or any other procedure for that matter, does not need them anymore, the `Controller` would call `release_context_id(<context_id>)` "signalling" the `Cleaner` that _it is now ok_ to cleanup any db artifacts with that `context_id` on the relevant nodes and do it **at its own pace**.

Now, the change on this pr is to drop the db artifacts, synchronously. 
Having a method named `cleanup_context_id` do what you propose would be misleading since what this method would actually do is much more complex. It will try to call the cleanup task on all the nodes and if then if there are nodes for which the cleanup task fail, for whatever reason, it will put the specified cleanup entry in a the cleanup loop until it succeeds. In order for someone to figure out all this, the code of the method has to be studied. This should not be the case. 
Instead 2 methods are defined with very specific scope, `cleanup_context_id` for calling the cleanup __now__ and `release_context_id` for doing it asynchronously. And it only takes 2 lines of code, instead of a convoluted logic elsewhere.  
```
if not self._cleaner.cleanup_context_id(context_id=context_id):
   self._cleaner.release_context_id(context_id=context_id)
```